oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resourceserver/authentication/AbstractOAuth2AccessTokenAuthenticationToken.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.resourceserver.authentication;
+
+import org.springframework.security.authentication.AbstractAuthenticationToken;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.SpringSecurityCoreVersion;
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Base class for {@link AbstractAuthenticationToken} implementations
+ * that expose common attributes between different OAuth 2.0 Access Token Formats.
+ *
+ * <p>
+ * For example, a {@link Jwt} could expose it's {@link Jwt#getClaims() claims} via
+ * {@link #getTokenAttributes()} or an &quot;Introspected&quot; OAuth 2.0 Access Token
+ * could expose the attributes of the Introspection Response via {@link #getTokenAttributes()}.
+ *
+ * @author Joe Grandja
+ * @since 5.1
+ * @see OAuth2AccessToken
+ * @see Jwt
+ * @see <a target="_blank" href="https://tools.ietf.org/search/rfc7662#section-2.2">2.2 Introspection Response</a>
+ */
+public abstract class AbstractOAuth2AccessTokenAuthenticationToken extends AbstractAuthenticationToken {
+	private static final long serialVersionUID = SpringSecurityCoreVersion.SERIAL_VERSION_UID;
+	private static final String DEFAULT_SCOPE_ATTRIBUTE_NAME = "scope";
+	private String scopeAttributeName = DEFAULT_SCOPE_ATTRIBUTE_NAME;
```

Comment:
So, we have `ClaimsAccessor` and `JwtClaimsAccessor`, which I think are helpful interfaces for providing a projection over a map of attributes. Here, this token is doing the same kind of projection for scopes. I wonder if there is a way to follow the same pattern with those accessors?

I realize that their is a hint of a naming convention problem with `ClaimsAccessor`, but it seems like it could go there or in a sub-interface.

Reply:
> I wonder if there is a way to follow the same pattern with those accessors?

Yes there is. I will apply that change to make use of `ClaimsAccessor` and ultimately simplify the internal method implementations. Great suggestion!